### ISSUE TITLE
Tendermint

### ISSUE THREAD

**Issue by faddat** [ponzu-cms/ponzu#127]
As you may or may not know, I'm working on https://github.com/dawn-network/glogchain, (and I'm looking at ponzu now to generate an API for glogchain that I'll connect to the [tendermint](https://tendermint.com/) stuff) which is a blockchain-based mediat platform.  Since blockchains are very, very computationally/storage/network expensive to operate, one of the key concepts with the chain I've been working on is to keep the "heavy" state entirely off-chain and store only IPFS hashes on the chain.  So maybe, one thing I am saying is that my chain may not be the perfect example of this.  

Anyway-- [tendermint](https://tendermint.com/) can fairly easily replace a web back end API with a distributed set of validator nodes, and it's pretty cool.  

Tendermint and their exchange cosmos benefit from the launch of new chains in fact-- all tendermint chains benefit when new tencdermint chains are launched in some way, because all the chains can talk to one another.  

So, basically, curious your thoughts on this.  I think block-chain back ends could fund a new generation of web systems that work for their users, instead of selling them out entirely.  

Please consider this a "tossing the idea out there"

and a 

"If this existed, would you use it?"

And if you know ponzu really really well--

"Any tips that might save me hours of head-banging-on-wall-feeling?

Thanks!

**Comment by nilslice** [comment]
Hey @faddat - I am familiar with your work with Dawn Project, very cool stuff. I have not yet seen the Tendermint project so I will take a look and get back to you about the rest of your questions.

**Comment by nilslice** [comment]
@faddat - I've attempted to learn a bit about tendermint and the [cosmos network](https://cosmos.network), and from what limited amount I understand, it seems interesting.. I'll admit, I have no experience in the world of blockchains and don't immediately see how Ponzu would be used in the hub / node / exchange diagram so to speak. 

If you would be willing / able to share more with me about what you'd like to do and how Ponzu might fit into that plan I'd happily provide some direction on how to get Ponzu to do what you want it to -- or advise you to use a different approach ;)

**Comment by faddat** [comment]
I'm willing and able :).

So here's a little bit that I've written about the end-state goals of our work:

https://quip.com/PKG6A4P7RMqt

But please know that we're very much a WIP.  Glogchain is designed to store state in a few different places.  

* tendermint (financial tx & pointers to content in IPFS, which will get delivered in IPFS or webtorrent or bittorrent depending on how the user is accessing gloghain).  We want to keep the chain as light as possible in order to keep things speedy.  


* distributed db - this is the hardest, trickiest part by far!  So Ponzu uses bolt, and that's the right way for it for sure. But now think that the end goal is to have many apps, all serving up one of those wonderful Ponzu'd APIs, from many servers.  I still haven't quite figured out if each machine should hold all of the state, or if the distributed database should distribute the state.  I prefer the distributed route but I haven't found anything that seems to fit this use case outside of -- possibly -- the javascript database called GunDB.  Noms looks decent, but it does store all of the state on all of the machines, which may not be ideal.  The distributed database should also keep things as light as possible.  I'm now wondering if anyone has come up with a way to link BoltDBs together in an eventually consistent or even sharded fashion.  


* IPFS - Actual content as markdown / mp4 / mp3

I love what Ponzu does in terms of giving all of the latest tech to its user in a simple and clean fashion.

**Comment by nilslice** [comment]
@faddat - I apologize for not replying to this.. it is great info and I need to look into all the projects you mention. I will, but I am in the midst of a relocation and have very little free time.. I have been doing more work recently with Ethereum smart contracts and think you are onto something interesting (based on how I understand your comments). I see now how Ponzu could offer a alternative backend, powered by blockchain (either public or private -- which opens up some interesting opportunities). 

Somewhat related, if you're available/interested, I'd like to discuss the possibility of building a fully open-source "company" around Ponzu, possibly raising some funds via an ICO.. I think going to the ICO market with an actual product and existing community would be well-received. 

I already have in the works some paid/premium business services like hosting, deeper API analytics, backups, and a few more products that would be very useful to Ponzu users. As the community and user-base grows, I believe a successful business could be build around it (much like Automattic around Wordpress).

Initially, I think there could be an ICO that would entitle owners to share of future profits generated by the premium services. The tokens/coins reserved by the "company" would additionally be an optional alternative to USD to pay developers who are interested in helping build needed features based on bounties for the jobs. The "company" would be a decentralized organization, and I'd be able to offer existing contributors coins for their work.

You're clearly quite involved and invested in the blockchain space, so it would be great to get your feedback if you have thoughts or ideas here.

**Comment by faddat** [comment]
I just underwent a move myself, and am looking to do same. How might we connect properly to put it into full gear?

**Comment by faddat** [comment]
I do have thoughts or ideas.  First off that isn't the usual manner for an ICO but you're right that this space is ready.  I have been moving as well.  I am going to fire off an e-mail to your boss sauce address, provided that I can find it.  

Talk to you soon,
-Jake

**Comment by nilslice** [comment]
@faddat - check your gitter notifications, we should be able to chat there. Or, email me at steve [at] bosssauce.it

Thanks!

**Comment by joeblew99** [comment]
Having read the requirements I think ponzu can do it. For scaling the system / sharing it out just need to use a basic choreography pattern.

Ponzu already has pre and post events, so you just need to stick NATS as a control plane so different ponzu shards can participate together in the choreography.

I might add that I really want to add materialised views to ponzu to make all the above easier and increase perf 
This will also allow global CDN like distribution. It's not a big addition either 

Happy to discuss further and try to get this distributed scaling into ponzu

**Comment by nilslice** [comment]
>Having read the requirements I think ponzu can do it. For scaling the system / sharing it out just need to use a basic choreography pattern.

This would be a fairly new concept to me, so I apologize if it takes a bit more explaining to me in order for me to fully grasp. Do you have any recommended resources to look into the basics of these patterns?

>Ponzu already has pre and post events, so you just need to stick NATS as a control plane so different ponzu shards can participate together in the choreography.

Sounds cool! I think we'd probably need to build in different hooks for this, since the default implementation on an `item.Item` is overridden when implemented by a Ponzu user. This, of course, is assuming you were referring to the `item.Hookable` interface with methods like `AfterSave`, etc. My understanding is that you'd propose we "echo" database actions throughout the network of chained ponzu's using these hook-like methods. 

>I might add that I really want to add materialised views to ponzu to make all the above easier and increase perf
This will also allow global CDN like distribution. It's not a big addition either

Yes, it's been something Ponzu has desperately lacked, and a direct result of there being no formal design of the Admin prior to me starting this project. However, now that the `system/db` package is pretty much set, creating the materialized views based on the existing routes in `system/admin/server.go`, it should be pretty straightforward. We'd need to update the UI to use templates (which really needs to be done anyway), since the code just inlines data with byte slices of UI.. :sweat_smile: 

Could you elaborate a bit more about the CDN-like distribution? You're referring to the distribution of synchronized ponzu instances correct? 

>Happy to discuss further and try to get this distributed scaling into ponzu

Great - should I start a "project" (like Trello board) on this repository? @faddat - would love to get your thoughts on this too. I think NATS, like @joeblew99 mentions, would be a good option for the synchronization between instances, but could Tenderment also achieve this or are they totally separate concepts?

**Comment by olliephillips** [comment]
Closing this issue. No activity. Please reopen if need to.
